File: app/src/main/java/org/odk/collect/android/formentry/FormEndView.java
package org.odk.collect.android.formentry;

import android.content.Context;
import android.text.Editable;
import android.text.InputFilter;
import android.text.TextWatcher;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.core.widget.NestedScrollView;

import org.odk.collect.android.R;
import org.odk.collect.android.listeners.SwipeHandler;
import org.odk.collect.android.utilities.FormNameUtils;

public class FormEndView extends SwipeHandler.View {

    private final Listener listener;
    private final String formTitle;
    private final String defaultInstanceName;

    public FormEndView(Context context, String formTitle, String defaultInstanceName, boolean instanceComplete, Listener listener) {
        super(context);
        this.formTitle = formTitle;
        this.defaultInstanceName = defaultInstanceName;
        this.listener = listener;
        init(context, instanceComplete);
    }

    private void init(Context context, boolean instanceComplete) {
        inflate(context, R.layout.form_entry_end, this);

        ((TextView) findViewById(R.id.description)).setText(context.getString(R.string.save_enter_data_description, formTitle));

        EditText saveAs = findViewById(R.id.save_name);

        // disallow carriage returns in the name
        InputFilter returnFilter = (source, start, end, dest, dstart, dend) -> FormNameUtils.normalizeFormName(source.toString().substring(start, end), true);
        saveAs.setFilters(new InputFilter[]{returnFilter});

        saveAs.setText(defaultInstanceName);
        saveAs.addTextChangedListener(new TextWatcher() {
            @Override
            public void afterTextChanged(Editable s) {
                listener.onSaveAsChanged(s.toString());
            }

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }
        });

        final CheckBox markAsFinalized = findViewById(R.id.mark_finished);
        markAsFinalized.setChecked(instanceComplete);

        findViewById(R.id.save_exit_button).setOnClickListener(v -> {
            listener.onSaveClicked(markAsFinalized.isChecked());
        });
    }

    @Override
    public boolean shouldSuppressFlingGesture() {
        return false;
    }

    @Nullable
    @Override
    public NestedScrollView getVerticalScrollView() {
        return findViewById(R.id.scroll_view);
    }

    public interface Listener {
        void onSaveAsChanged(String string);

        void onSaveClicked(boolean markAsFinalized);
    }
}

File: collect_app/src/main/java/org/odk/collect/android/formentry/FormEndView.kt
package org.odk.collect.android.formentry

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import androidx.core.view.isVisible
import androidx.core.widget.NestedScrollView
import org.odk.collect.android.R
import org.odk.collect.android.databinding.FormEntryEndBinding

class FormEndView(
    context: Context,
    formTitle: String,
    isFormEditableAfterFinalization: Boolean,
    formEndViewModel: FormEndViewModel,
    private val listener: Listener
) : SwipeHandler.View(context) {

    private val binding = FormEntryEndBinding.inflate(LayoutInflater.from(context), this, true)

    init {
        binding.description.text = context.getString(org.odk.collect.strings.R.string.save_enter_data_description, formTitle)

        binding.saveAsDraft.isVisible = formEndViewModel.isSaveDraftEnabled()
        binding.saveAsDraft.setOnClickListener {
            listener.onSaveClicked(false)
        }

        binding.finalize.isVisible = formEndViewModel.isFinalizeEnabled()
        binding.finalize.setOnClickListener {
            listener.onSaveClicked(true)
        }

        binding.divider.isVisible = binding.saveAsDraft.isVisible && binding.finalize.isVisible

        val shouldFormBeSentAutomatically = formEndViewModel.shouldFormBeSentAutomatically()
        if (shouldFormBeSentAutomatically) {
            binding.finalize.text = context.getString(org.odk.collect.strings.R.string.send)
        }

        if (binding.saveAsDraft.isVisible && binding.finalize.isVisible) {
            if (shouldFormBeSentAutomatically) {
                if (isFormEditableAfterFinalization) {
                    setWarning(
                        icon = R.drawable.ic_edit_24,
                        title = org.odk.collect.strings.R.string.form_editing_enabled_after_sending,
                        hint = org.odk.collect.strings.R.string.form_editing_enabled_after_sending_hint
                    )
                } else {
                    setWarning(
                        icon = R.drawable.ic_edit_off_24,
                        title = org.odk.collect.strings.R.string.form_editing_disabled_after_sending,
                        hint = org.odk.collect.strings.R.string.form_editing_disabled_hint
                    )
                }
            } else {
                if (isFormEditableAfterFinalization) {
                    setWarning(
                        icon = R.drawable.ic_edit_24,
                        title = org.odk.collect.strings.R.string.form_editing_enabled_after_finalizing,
                        hint = org.odk.collect.strings.R.string.form_editing_enabled_after_finalizing_hint
                    )
                } else {
                    setWarning(
                        icon = R.drawable.ic_edit_off_24,
                        title = org.odk.collect.strings.R.string.form_editing_disabled_after_finalizing,
                        hint = org.odk.collect.strings.R.string.form_editing_disabled_hint
                    )
                }
            }
        } else if (binding.finalize.isVisible) {
            if (shouldFormBeSentAutomatically) {
                if (isFormEditableAfterFinalization) {
                    setWarning(
                        icon = R.drawable.ic_edit_24,
                        title = org.odk.collect.strings.R.string.form_editing_enabled_after_sending,
                        hint = org.odk.collect.strings.R.string.form_editing_enabled_after_sending_hint
                    )
                } else {
                    setWarning(
                        icon = R.drawable.ic_edit_off_24,
                        title = org.odk.collect.strings.R.string.form_editing_disabled_after_sending,
                        hint = null
                    )
                }
            } else {
                if (isFormEditableAfterFinalization) {
                    setWarning(
                        icon = R.drawable.ic_edit_24,
                        title = org.odk.collect.strings.R.string.form_editing_enabled_after_finalizing,
                        hint = org.odk.collect.strings.R.string.form_editing_enabled_after_finalizing_hint
                    )
                } else {
                    setWarning(
                        icon = R.drawable.ic_edit_off_24,
                        title = org.odk.collect.strings.R.string.form_editing_disabled_after_finalizing,
                        hint = null
                    )
                }
            }
        } else {
            binding.formEditsWarning.visibility = View.GONE
        }
    }

    private fun setWarning(icon: Int, title: Int, hint: Int?) {
        binding.formEditsIcon.setImageResource(icon)
        binding.formEditsWarningTitle.setText(title)

        if (hint != null) {
            binding.formEditsWarningMessage.setText(hint)
        }
    }

    override fun shouldSuppressFlingGesture() = false

    override fun verticalScrollView(): NestedScrollView? {
        return findViewById(R.id.scroll_view)
    }

    interface Listener {
        fun onSaveClicked(markAsFinalized: Boolean)
    }
}
